demos/console.php

```diff
@@ -8,28 +8,80 @@ class Test extends \atk4\data\Model
+        $this->debug('debug {foo}', ['foo'=>'bar']);
+        $this->emergency('emergency {foo}', ['foo'=>'bar']);
+        $this->alert('alert {foo}', ['foo'=>'bar']);
+        $this->critical('critical {foo}', ['foo'=>'bar']);
+        $this->error('error {foo}', ['foo'=>'bar']);
+        $this->warning('warning {foo}', ['foo'=>'bar']);
+        $this->notice('notice {foo}', ['foo'=>'bar']);
+        $this->info('info {foo}', ['foo'=>'bar']);
 
         return 123;
     }
 }
 
-$app->add('Console')->set(function ($console) {
+$tt = $app->add('Tabs');
+
+$t = $tt->addTab('set()');
+$t->add([
+    'Header',
+    'icon'=> 'terminal',
+    'Console output streaming',
+    'subHeader'=> 'any output your PHP script produces through console is displayed to user in real-time',
+]);
+$t->add('Console')->set(function ($console) {
     $console->output('Executing test process...');
     sleep(1);
     $console->output('Now trying something dangerous..');
     sleep(1);
+    echo 'direct output is captured';
+
+    throw new \atk4\data\Exception('BOOM - exceptions are caught');
+});
 
-    throw new \atk4\data\Exception('BOOM');
-    $console->output('hello there');
+$t = $tt->addTab('runMethod()', function ($t) {
+    $t->add([
+        'Header',
+        'icon'=> 'terminal',
+        'Non-interractive method invocation',
+        'subHeader'=> 'console can invoke a method, which normaly would be non-interractive and can still capture debug output',
+    ]);
+    $t->add('Console')->runMethod($t->add(new Test()), 'generateReport');
+});
 
-    //$console->runCommand('ls', ['/etc']);
+$t = $tt->addTab('exec() single', function ($t) {
+    $t->add([
+        'Header',
+        'icon'=> 'terminal',
+        'Command execution',
+        'subHeader'=> 'it is easy to run server-side commands and stream output through console',
+    ]);
+    $t->add('Console')->exec('/bin/pwd');
+});
 
-    //$console->send(new \atk4\ui\jsExpression('alert(1)'));
+$t = $tt->addTab('exec() chain', function ($t) {
+    $t->add([
+        'Header',
+        'icon'=> 'terminal',
+        'Command execution',
+        'subHeader'=> 'it is easy to run server-side commands and stream output through console',
+    ]);
+    $t->add('Console')->set(function ($c) {
+        $c
+            ->exec('/sbin/ping', ['-c', '5', '-i', '1', '192.168.0.1'])
+            ->exec('/sbin/ping', ['-c', '5', '-i', '2', '8.8.8.8'])
+            ->exec('/bin/no-such-command');
+    });
 });
 
-$app->add('Console')->setModel($app->add(new Test()), 'generateReport');
+$t = $tt->addTab('composer update', function ($t) {
+    $t->add([
+        'Header',
+        'icon'=> 'terminal',
+        'Command execution',
+        'subHeader'=> 'it is easy to run server-side commands and stream output through console',
+    ]);
+    $t->add('Console')->exec('bash', ['-c', 'cd ..; echo "Running \'composer update\' in `pwd`"; composer --no-ansi update; echo "Self-updated. OK to refresh now!"']);
```

Comment:
Doesn't work on Windows. Please try to stick to command which also works on windows in demo.

Reply:
oh and ping worked? 

perhaps just disable this demo on windows?

Reply:
not sure how to disable it in windows from php side.